### PR TITLE
G-API: Support OpenVINO Execution Provider for ONNXRT Backend

### PR DESCRIPTION
**Pull request by TolyaTalamanov**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [ ] I agree to contribute to the project under Apache 2 License.
- [ ] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [ ] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/gapi/include/opencv2/gapi/infer/onnx.hpp [hunk 1]
```diff
@@ -27,6 +27,57 @@ namespace gapi {
  */
 namespace onnx {
 
+namespace ep {
+
+struct GAPI_EXPORTS_W_SIMPLE OpenVINO {
```
Comment (TolyaTalamanov) [inline modules/gapi/include/opencv2/gapi/infer/onnx.hpp:32]: requires doxygen string

File: modules/gapi/include/opencv2/gapi/infer/onnx.hpp [hunk 2]
```diff
@@ -27,6 +27,57 @@ namespace gapi {
...
+    GAPI_WRAP
+    OpenVINO() = default;
```
Comment (TolyaTalamanov) [inline modules/gapi/include/opencv2/gapi/infer/onnx.hpp:34]: Only needed for python, worth to comment

File: modules/gapi/include/opencv2/gapi/infer/onnx.hpp [hunk 3]
```diff
@@ -27,6 +27,57 @@ namespace gapi {
...
+
+    GAPI_WRAP
+    OpenVINO(const std::string &id)
+        : device_id(id) {
+    }
+
+    GAPI_WRAP
+    OpenVINO& cfgDeviceType(const std::string &type) {
```
Comment (TolyaTalamanov) [inline modules/gapi/include/opencv2/gapi/infer/onnx.hpp:42]: add doxygen strings for all methods

File: modules/gapi/include/opencv2/gapi/infer/onnx.hpp [hunk 4]
```diff
@@ -279,6 +334,16 @@ template<typename Net> class Params {
         return *this;
     }
 
+    Params<Net>& cfgExecutionProvider(ep::OpenVINO&& ov_ep) {
```
Comment (TolyaTalamanov) [inline modules/gapi/include/opencv2/gapi/infer/onnx.hpp:337]: doxygen string

File: modules/gapi/src/backends/onnx/gonnxbackend.cpp [hunk 5]
```diff
@@ -143,6 +143,39 @@ class ONNXCompiled {
     void run();
 };
 
+void appendExecutionProvider(Ort::SessionOptions          *session_options,
+                             const cv::gapi::onnx::ep::EP &execution_provider) {
+    namespace ep = cv::gapi::onnx::ep;
+    switch (execution_provider.index()) {
+        case ep::EP::index_of<ep::OpenVINO>(): {
```
Comment (TolyaTalamanov) [inline modules/gapi/src/backends/onnx/gonnxbackend.cpp:150]: maybe some log info which exactly EP is selected would be usefule

File: modules/gapi/src/backends/onnx/gonnxbackend.cpp [hunk 6]
```diff
@@ -143,6 +143,39 @@ class ONNXCompiled {
...
+                options.device_type = ovep.device_type->c_str();
+             }
+             if (ovep.cache_dir) {
+                options.cache_dir = ovep.cache_dir->c_str();
+             }
+             if (ovep.num_of_threads) {
+                options.num_of_threads = *ovep.num_of_threads;
+             }
+
+             session_options->AppendExecutionProvider_OpenVINO(options);
```
Comment (TolyaTalamanov) [inline modules/gapi/src/backends/onnx/gonnxbackend.cpp:171]: This thing might throw, make sense to wrap it and show the understandable error message...

File: modules/gapi/src/backends/onnx/gonnxbackend.cpp [hunk 7]
```diff
@@ -143,6 +143,39 @@ class ONNXCompiled {
...
+             OrtOpenVINOProviderOptions options;
+
+             options.device_id = ovep.device_id.c_str();
+
+             options.enable_vpu_fast_compile = ovep.enable_vpu_fast_compile;
+             options.context = ovep.context;
+             options.enable_opencl_throttling = ovep.enable_opencl_throttling;
+             options.enable_dynamic_shapes = ovep.enable_dynamic_shapes;
+
+             if (ovep.device_type) {
```
Comment (TolyaTalamanov) [inline modules/gapi/src/backends/onnx/gonnxbackend.cpp:161]: Put some explanation why some of the parameters is `optional`

File: modules/gapi/include/opencv2/gapi/infer/onnx.hpp (hunk 1 above)
Comment (TolyaTalamanov) [reply modules/gapi/include/opencv2/gapi/infer/onnx.hpp:40]: Done

File: modules/gapi/include/opencv2/gapi/infer/onnx.hpp (hunk 2 above)
Comment (TolyaTalamanov) [reply modules/gapi/include/opencv2/gapi/infer/onnx.hpp:44]: Done

File: modules/gapi/include/opencv2/gapi/infer/onnx.hpp (hunk 3 above)
Comment (TolyaTalamanov) [reply modules/gapi/include/opencv2/gapi/infer/onnx.hpp:67]: Done

File: modules/gapi/src/backends/onnx/gonnxbackend.cpp (hunk 7 above)
Comment (TolyaTalamanov) [reply modules/gapi/src/backends/onnx/gonnxbackend.cpp:158]: Done

File: modules/gapi/src/backends/onnx/gonnxbackend.cpp (hunk 5 above)
Comment (TolyaTalamanov) [reply modules/gapi/src/backends/onnx/gonnxbackend.cpp:150]: Done

File: modules/gapi/include/opencv2/gapi/infer/onnx.hpp (hunk 4 above)
Comment (TolyaTalamanov) [reply modules/gapi/include/opencv2/gapi/infer/onnx.hpp]: Done

**Comment by TolyaTalamanov** [comment]
@dmatveev Could you have a look, please?

**Comment by TolyaTalamanov** [comment]
The open question is how it can be tested because it requires `onnxruntime` to be built with `--use_openvino` that enables OpenVINO Execution Provider otherwise `AppendExecutionProvider_OpenVINO` will fail with exception

File: modules/gapi/src/backends/onnx/gonnxbackend.cpp [hunk 8]
```diff
@@ -143,6 +143,41 @@ class ONNXCompiled {
     void run();
 };
 
+static void appendExecutionProvider(Ort::SessionOptions          *session_options,
+                                    const cv::gapi::onnx::ep::EP &execution_provider) {
+    namespace ep = cv::gapi::onnx::ep;
+    switch (execution_provider.index()) {
+        case ep::EP::index_of<ep::OpenVINO>(): {
+             GAPI_LOG_INFO(NULL, "OpenVINO Execution Provider is selected.");
```
Comment (dmatveev) [inline modules/gapi/src/backends/onnx/gonnxbackend.cpp:151]: Maybe it makes sense to display for which network (if it is not mentioned in the log earlier).

File: modules/gapi/src/backends/onnx/gonnxbackend.cpp [hunk 9]
```diff
@@ -143,6 +143,41 @@ class ONNXCompiled {
...
+                 std::stringstream ss;
+                 ss << "ONNX Backend: Failed to enable OpenVINO Execution Provider: "
+                    << e.what() << "\nMake sure that onnxruntime has"
+                                   " been compiled with OpenVINO support.";
+                 cv::util::throw_error(std::runtime_error(ss.str()));
+             }
+             break;
+        }
+        default:
+            break;
```
Comment (dmatveev) [inline modules/gapi/src/backends/onnx/gonnxbackend.cpp:177]: Any message for this case?

**Review by dmatveev — APPROVED** [review]
`Generally looks good

File: modules/gapi/src/backends/onnx/gonnxbackend.cpp [hunk 10]
```diff
@@ -143,6 +143,41 @@ class ONNXCompiled {
     void run();
 };
 
+static void appendExecutionProvider(Ort::SessionOptions          *session_options,
```
[the remark this answers is not in the document]
Comment (TolyaTalamanov) [reply modules/gapi/src/backends/onnx/gonnxbackend.cpp:146]: hmm, you actually make a good point, what if it's possible to use multiple execution providers at the same time...
In that case I'd probably need to hold the list of them.

File: modules/gapi/src/backends/onnx/gonnxbackend.cpp (hunk 8 above)
Comment (TolyaTalamanov) [reply modules/gapi/src/backends/onnx/gonnxbackend.cpp:151]: As an option, yes

File: modules/gapi/src/backends/onnx/gonnxbackend.cpp (hunk 9 above)
Comment (TolyaTalamanov) [reply modules/gapi/src/backends/onnx/gonnxbackend.cpp:177]: I think some sort of log would be useful

File: modules/gapi/src/backends/onnx/gonnxbackend.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (TolyaTalamanov) [reply modules/gapi/src/backends/onnx/gonnxbackend.cpp:146]: Looks so https://onnxruntime.ai/docs/execution-providers/#use-execution-providers

**Comment by asmorkalov** [comment]
Tested manually with openvino_toolkit_ubuntu18_2022.1.1.7030.39aba80957e.

**Pull request merged**